pkg/usagestats/reporter_test.go

```diff
@@ -0,0 +1,300 @@
+
+			err := reporter.sendReport(context.Background(), buildReport(newClusterSeed(), time.Now(), time.Hour))
+			require.Error(t, err)
+			require.Contains(t, err.Error(), "failed to send the report to the stats server")
+			require.False(t, serverInvoked.Load())
+
+			require.NoError(t, testutil.GatherAndCompare(reg, strings.NewReader(`
+				# HELP cortex_usage_stats_report_sends_total The total number of attempted send requests.
+				# TYPE cortex_usage_stats_report_sends_total counter
+				cortex_usage_stats_report_sends_total 1
+
+				# HELP cortex_usage_stats_report_sends_failed_total The total number of failed send requests.
+				# TYPE cortex_usage_stats_report_sends_failed_total counter
+				cortex_usage_stats_report_sends_failed_total 1
+			`), "cortex_usage_stats_report_sends_total", "cortex_usage_stats_report_sends_failed_total"))
+		},
+	}
+
+	for testName, testRun := range tests {
+		t.Run(testName, func(t *testing.T) {
+			// Reset.
+			serverInvoked.Store(false)
+
+			testRun(t)
+		})
+	}
+}
+
+func TestReporter_SendReportPeriodically(t *testing.T) {
+	tests := map[string]struct {
+		failOneRequestEvery int
+	}{
+		"all requests to stats server succeed": {
+			failOneRequestEvery: 0,
+		},
+		"some requests to stats server fail and they get retried": {
+			failOneRequestEvery: 2,
+		},
+	}
+
+	for testName, testData := range tests {
+		testData := testData
+
+		t.Run(testName, func(t *testing.T) {
+			t.Parallel()
+
+			var (
+				ctx           = context.Background()
+				bucketClient  = prepareLocalBucketClient(t)
+				reportsMx     sync.Mutex
+				reports       []Report
+				requestsCount = atomic.NewInt32(0)
+			)
+
+			// Mock the stats server.
+			server := httptest.NewServer(http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
+				requestsCount.Inc()
+				if testData.failOneRequestEvery > 0 && requestsCount.Load()%int32(testData.failOneRequestEvery) == 0 {
+					writer.WriteHeader(http.StatusServiceUnavailable)
+					return
+				}
+
+				data, err := io.ReadAll(request.Body)
+				require.NoError(t, err)
+
+				report := Report{}
+				require.NoError(t, json.Unmarshal(data, &report))
+
+				reportsMx.Lock()
+				reports = append(reports, report)
+				reportsMx.Unlock()
+
+				writer.WriteHeader(http.StatusOK)
+			}))
+			defer server.Close()
+
+			r := NewReporter(bucketClient, log.NewNopLogger(), prometheus.NewPedanticRegistry())
+			r.serverURL = server.URL
+			r.reportCheckInterval = 100 * time.Millisecond
+			r.reportSendInterval = time.Second
+
+			// Upload the seed file.
+			seed := newClusterSeed()
+			seed.CreatedAt = time.Now().Add(-time.Hour)
+			require.NoError(t, writeSeedFile(ctx, r.bucket, seed))
+
+			// Start the reporter.
+			require.NoError(t, services.StartAndAwaitRunning(ctx, r))
+			t.Cleanup(func() {
+				require.NoError(t, services.StopAndAwaitTerminated(ctx, r))
+			})
+
+			// Wait some time and check the received reports.
+			time.Sleep(6 * time.Second)
```

Comment:
I guess we can live with this assuming that golang executes tests from multiple packages in parallel and there are already tests longer than 6s.

Reply:
Actually I will improve it and use the "wait condition" utility we have for tests, with a longer timeout.